clang/test/dpct/helper_files_ref/include/sparse_utils.hpp

```diff
@@ -40,6 +41,216 @@ class matrix_info {
+/// \param [in] row_ptr An array of length \p num_rows + 1.
+/// \param [in] col_ind An array containing the column indices in index-based
+/// numbering.
+/// \param [in] b Data of the matrix B.
+/// \param [in] ldb Leading dimension of the matrix B.
+/// \param [in] beta Scaling factor for the matrix B.
+/// \param [in, out] c Data of the matrix C.
+/// \param [in] ldc Leading dimension of the matrix C.
+template <typename T>
+void csrmm(sycl::queue &queue, oneapi::mkl::transpose trans, int sparse_rows,
+           int dense_cols, int sparse_cols, const T *alpha,
+           const std::shared_ptr<matrix_info> info, const T *val,
+           const int *row_ptr, const int *col_ind, const T *b, int ldb,
+           const T *beta, T *c, int ldc) {
+  using Ty = typename dpct::DataType<T>::T2;
+  //static_assert(
+  //    std::disjunction_v<std::is_same<Ty, float>, std::is_same<Ty, double>>,
+  //    "Unimplemented functionality: oneapi::mkl::sparse::gemm "
+  //    "currently does not support complex data types");
+  auto alpha_value =
+      detail::get_value(reinterpret_cast<const Ty *>(alpha), queue);
+  auto beta_value =
+      detail::get_value(reinterpret_cast<const Ty *>(beta), queue);
+
+  oneapi::mkl::sparse::matrix_handle_t *sparse_matrix_handle =
+      new oneapi::mkl::sparse::matrix_handle_t;
+  oneapi::mkl::sparse::init_matrix_handle(sparse_matrix_handle);
+  auto data_row_ptr = detail::get_memory(const_cast<int *>(row_ptr));
+  auto data_col_ind = detail::get_memory(const_cast<int *>(col_ind));
+  auto data_val =
+      detail::get_memory(reinterpret_cast<Ty *>(const_cast<T *>(val)));
+  oneapi::mkl::sparse::set_csr_data(queue, *sparse_matrix_handle, sparse_rows,
+                                    sparse_cols, info->get_index_base(),
+                                    data_row_ptr, data_col_ind, data_val);
+
+  auto data_b = detail::get_memory(reinterpret_cast<Ty *>(const_cast<T *>(b)));
+  auto data_c = detail::get_memory(reinterpret_cast<Ty *>(c));
+  switch (info->get_matrix_type()) {
+  case matrix_info::matrix_type::ge: {
+    oneapi::mkl::sparse::gemm(queue, oneapi::mkl::layout::row_major, trans,
+                              oneapi::mkl::transpose::nontrans, alpha_value,
+                              *sparse_matrix_handle, data_b, dense_cols, ldb,
+                              beta_value, data_c, ldc);
+    break;
+  }
+  default:
+    throw std::runtime_error(
+        "the csrmm does not support matrix_info::matrix_type::sy, "
+        "matrix_info::matrix_type::tr amd matrix_info::matrix_type::he");
+  }
+
+  sycl::event e =
+      oneapi::mkl::sparse::release_matrix_handle(queue, sparse_matrix_handle);
+  queue.submit([&](sycl::handler &cgh) {
+    cgh.depends_on(e);
+    cgh.host_task([=] { delete sparse_matrix_handle; });
+  });
+}
+
+/// Saving the optimization information for solving a system of linear
+/// equations.
+class optimize_info {
+public:
+  /// Constructor
```

Comment:
ctor: new
dtor: host_task depends on release, delete
create: init
destroy: release

Reply:
I changed the dtor to a sycn-api.